### ISSUE TITLE
Importance of inputs and positive weights

### ISSUE THREAD

**Issue by AndZark** [fchollet/keras#6557]
Greetings,

I have 2 questions:

1) How can one find, using Keras, which dimension of the input is the most important?

2) Is there a way to restrict the weights to be positive?

Thank you for your time!

**Comment by bstriner** [comment]
1) Way ill-defined. Do you mean PCA or smth? You could also mean the largest gradient of output to input. How do you define important?

2) You can make a constraint, or just regularize so things are positive. Read up the docs on constraints and regularization. Simplest example would be adding `kernel_constraint=K.relu` to your layer constructor.

Cheers

**Comment by AndZark** [comment]
bstriner thank you for your answer!

I will explain the situation:

I have figures for 9 days (input with 9 dimensions) and I wish to predict the figure for the 10-th day (this is the output). Now the figure of the 9th day is the most indicative for the prediction. That's what i mean by "most important dimension" and I was wondering if there is a way I can see that using Keras.

Cheers

**Comment by bstriner** [comment]
"Most indicative" is still a squishy concept. One way to define it is to make multiple models dropping out one of the inputs. So you train a model given all days but day 1, train a model given all days but day 2, etc. Also train your main model given all 9 days. Compare the loss between models. If dropping out day 9 increases the loss by the most compared to using all of the days, then day 9 was the most important.

In this example, I am defining "most important dimension" as "dimension that increases the loss the most when removed". This is not the only way to define it. Machine learning is all about operationalizing squishy things like "most important".

Alternatively, you could train a model given just day 1, just day 2, etc., and see which model has the lowest loss. Here, "most important dimension" is "dimension that individually can be used for the best prediction".

Depending on how you define it you will get different answers.

Cheers

**Comment by AndZark** [comment]
Thank you a lot for your suggestions bstriner!

I will leave the thread open for a few more days and if no one else answers I will close it.

Cheers
